File: g1/XZ/plotym.py
import concore
import numpy as np
import matplotlib.pyplot as plt
import time

print("plotym - Live Mode")

# Simulation Configuration
concore.delay = 0.02
concore.default_maxtime(150)
init_simtime_ym = "[0.0, 0.0]"
ym = concore.initval(init_simtime_ym) #

# Initialize Live Plot
plt.ion()  # Turn on interactive mode
fig, ax = plt.subplots()
line1, = ax.plot([], [], 'b-', label='ym')
ax.set_ylabel('ym')
ax.set_xlabel('Cycles')
ax.legend(loc='upper right')

# Data buffers for live plotting
x_data = []
ym1_data = []

# Ensure the plot window is visible before starting
plt.show()

try:
    while(concore.simtime < concore.maxtime): #
        while concore.unchanged(): #
            ym = concore.read(1, "ym", init_simtime_ym) #
        
        concore.write(1, "ym", ym) #
        print("ym=" + str(ym))
        
        # Update data buffers
        current_step = len(x_data)
        x_data.append(current_step)
        ym1_data.append(ym[0])
        
        # Update plot line data
        line1.set_data(x_data, ym1_data)
        
        # Adjust axes limits dynamically
        ax.relim()
        ax.autoscale_view()
        
        # Pause briefly to allow the GUI thread to refresh the window
        plt.pause(0.001)

except KeyboardInterrupt:
    print("\nSimulation interrupted by user.")

print("retry=" + str(concore.retrycount)) #

# Finalize plot
plt.ioff() # Turn off interactive mode
plt.savefig("ym_live_final.pdf")
plt.show() # Keep plot open after simulation ends

File: example/plotym.jl
import Concore: initval, concore_read, concore_write, unchanged, state, default_maxtime
using Plots

println("plotym - Live Mode")

# Simulation Configuration
state.delay = 0.02
default_maxtime(150.0)
const init_simtime_ym = "[0.0, 0.0]"

global ym = initval(init_simtime_ym)

# Initialize Live Plot
plt = plot(Float64[], Float64[], title="Live plots", xlabel="Cycles", ylabel="ym", legend=:topright, label="ym", lw=2)

x_data = Int[]
ym1_data = Float64[]

try
    while state.simtime < state.maxtime
        while unchanged()
            global ym = concore_read(1, "ym", init_simtime_ym)
        end
        
        concore_write(1, "ym", ym, 0.0)
        println("ym=$ym")
        
        push!(x_data, length(x_data))
        push!(ym1_data, ym[1])
        
        # Update plot line data by pushing newest point
        push!(plt, 1, x_data[end], ym1_data[end])
        
        sleep(0.001)
    end
catch e
    if isa(e, InterruptException)
        println("\nSimulation interrupted by user.")
    else
        rethrow(e)
    end
end

println("retry=$(state.retrycount)")

# Finalize plot
savefig(plt, "ym_live_final.pdf")
